pkg/resource/autonaming/pattern.go

```diff
@@ -0,0 +1,183 @@
+		if err != nil {
+			configErr = err
+			return ""
+		}
+		return v
+	})
+	if configErr != nil {
+		return "", configErr
+	}
+	return pattern, nil
+}
+
+// generateName generates a final proposed name based on the configured pattern, the resource URN, and random seed.
+// Note that the pattern is expected to have already had stack-level expressions like ${org}, ${project}, ${stack},
+// and ${config.key} resolved before passing to this function.
+func generateName(pattern string, urn urn.URN, randomSeed []byte) (string, bool) {
+	// Replace ${name} with the logical name
+	result := strings.ReplaceAll(pattern, "${name}", urn.Name())
+	hasRandom := false
+
+	// Create a random number generator with the given seed. If no seed is provided, use a default
+	// random number generator.
+	var random *frand.RNG
+	if len(randomSeed) == 0 {
+		random = frand.New()
+	} else {
+		// frand.NewCustom needs a 32 byte seed. Take the SHA256 hash of whatever bytes we've been given as a
+		// seed and pass the 32 byte result of that to frand.
+		hash := crypto.SHA256.New()
+		hash.Write(randomSeed)
+		seed := hash.Sum(nil)
+		bufsize := 1024 // Same bufsize as used by frand.New.
+		rounds := 12    // Same rounds as used by frand.New.
+		random = frand.NewCustom(seed, bufsize, rounds)
+	}
+
+	// Replace ${hex(n)} with random hex string of length n
+	hexRegex := regexp.MustCompile(`\${hex\((\d+)\)}`)
+	result = hexRegex.ReplaceAllStringFunc(result, func(match string) string {
+		n, _ := strconv.Atoi(hexRegex.FindStringSubmatch(match)[1])
+		b := make([]byte, n/2+1)
+		_, _ = random.Read(b)
+		hasRandom = true
+		return hex.EncodeToString(b)[:n]
+	})
+
+	// Replace ${alphanum(n)} with random alphanumeric string of length n
+	alphaRegex := regexp.MustCompile(`\${alphanum\((\d+)\)}`)
+	result = alphaRegex.ReplaceAllStringFunc(result, func(match string) string {
+		n, _ := strconv.Atoi(alphaRegex.FindStringSubmatch(match)[1])
+		const chars = "abcdefghijklmnopqrstuvwxyz0123456789"
+		b := make([]byte, n)
+		for i := range b {
+			b[i] = chars[random.Intn(len(chars))]
+		}
+		hasRandom = true
+		return string(b)
+	})
+
+	// Replace ${string(n)} with random letter string of length n
+	strRegex := regexp.MustCompile(`\${string\((\d+)\)}`)
+	result = strRegex.ReplaceAllStringFunc(result, func(match string) string {
+		n, _ := strconv.Atoi(strRegex.FindStringSubmatch(match)[1])
+		const chars = "abcdefghijklmnopqrstuvwxyz"
+		b := make([]byte, n)
+		for i := range b {
+			b[i] = chars[random.Intn(len(chars))]
+		}
+		hasRandom = true
+		return string(b)
+	})
+
+	// Replace ${num(n)} with random digit string of length n
+	numRegex := regexp.MustCompile(`\${num\((\d+)\)}`)
+	result = numRegex.ReplaceAllStringFunc(result, func(match string) string {
+		n, _ := strconv.Atoi(numRegex.FindStringSubmatch(match)[1])
+		const chars = "0123456789"
+		b := make([]byte, n)
+		for i := range b {
+			b[i] = chars[random.Intn(len(chars))]
+		}
+		hasRandom = true
+		return string(b)
+	})
```

Comment:
Might be worth breaking these out into helpers? `replaceString`, `replaceNum` etc? Not super fussed but could make it more obvious what we support/how to extend at a glance?